### ISSUE TITLE
Allow mutations on data

### ISSUE THREAD

**Issue by C-Duv** [google/mtail#162]
TL;DR: Create new fields from other fields (or replace values of existing ones) via regex or other builtin fonctions, just like `logstash`'s `mutate` plugin.

Context: I've just dicsovered this tool (and `grok_exporter`) after trying to perform tail+parse+count data processing in existing PHP application (disclaimer: it failed).

The ability to alter fields/labels before exporting to Prometheus (just like `logstash`'s `mutate` plugin) would be awesome.

In my use case I am reading Apache access.log file and I want to export HTTP requests count with the following dimensions/labels:
* Straightforward (simple named regex capture group):
  * Status code
  * HTTP verb
  * Response size
* Add fixed fields/labels (eg. input file name)
* Field computation required:
  * Only keep the base URL (scheme+FQDN) from referrer (eg. `http://example.com/foo.asp?id=42` => `http://example.com`): some regex would do (or I could adapt the _COND_ matching regex)
  * Extract query string part from the referrer and add some labels from it. For example, from `http://example.com/foo.asp?id=42&source=github&foo=bar` I want the following fields: `id` and `foo`. I get that dropping labels is already something `grok_exporter` can do, so having a mutation that creates a label for every found query parameter is fine.

Other use cases (not mine):
* Compute hash of string.
* Anonymize values (eg. for the the auth HTTP field, replace any value different than `-` by `connected_user` and the `-` value by `guest`).
* Replace raw values by meaningful values (eg. *404* => *Not Found*).

**Comment by jaqx0r** [comment]
Matching against other variables is tracked in #103 I think, which is mostly complete now, so try it out and let me know what doesn't work.

 The "straightforward" point should already be supported.
The "input file name" is a dupe of #36 , but you can already use getfilename() builtin: https://github.com/google/mtail/blob/master/docs/Language.md#builtin-functions

Field computation is new; we don't hvae any substring spuport, but I think this could be done with nested matches to extract based on regular expression if you specify each match, but I think you're asking for a way to match `?id=42&source=github&foo=bar` and have it create $id, $source, and $foo automatically?

With the exception of the hash computation, the other use cases should already be possible but if not please show me some mtail program and example input where it's not working.

**Comment by C-Duv** [comment]
I do indeed want to automatically create the `id`, `source` and `foo` fields from a field containing `?id=42&source=github&foo=bar`.

Right now I am using `logstash` with multiple `grok` and `kv` (+ `mutate` for cleaning):

```
input {
    stdin { } #DEBUG: For dummy run with fake data
}
filter {
    mutate { #DEBUG: For dummy run with fake data
        add_field => { "message" => '1.2.3.4 - - [10/Jun/2018:11:22:33 +0000] "GET /foo.asp?id=42&source=github&foo=bar HTTP/1.1" 200 666 "http://www.example.com" "curl"' }
    }

    # Match an Apache "combined" access.log line of log:
    grok {
        match => { "message" => "%{HTTPD_COMBINEDLOG}" }
        add_field => { "fixed_field" => "fixed_value" }
    }

    # Split HTTP requested URI into path and query parts:
    grok {
        match => { "request" => "%{URIPATH:request_url_path}(?:%{URIPARAM:request_url_query})?" }
    }

    # Keep only the main URL from referrer:
    grok {
        match => { "referrer" => "(?<referrer_main_url>%{URIPROTO}://(?:%{USER}(?::[^@]*)?@)?(?:%{URIHOST})?)(?:%{URIPATHPARAM})?" }
    }

    # Extract fields from URL's query part:
    kv {
        source => "request_url_query"
        field_split => "&?"
        value_split => "=" # (This is default)
        transform_key => "lowercase"
        transform_value => "lowercase"
    }

    mutate {
        # Remove any unwanted field
        remove_field => [ "clientip", "ident", "auth", "timestamp", "httpversion", "agent", "referrer" ]

        convert => [ "response", "integer" ]
        convert => [ "bytes", "integer" ]
    }
}
output {
    stdout {
        codec => rubydebug
    }
}
```

And I pass `logstash`'s output (using `line` codec instead of `rubydebug`) to `mtail`.

But, now I am wondering if asking such feature is not putting to much intelligence into `mtail` where a `logstash` works...

**Comment by jaqx0r** [comment]
Yep you're correct, I don't want to put specialised extraction methods in
mtail.  I am wondering though if we can create something with the existing
functionality.

**Comment by jaqx0r** [comment]
I was wondering if we could construct something like

/(&|?)(?P<key>.*)=(?P<value>.*)(&|$)/ {
  uri[$key] = $value
}

but I don't think it would work because the expression needs to be matched N times where N is the number of keys in the pattern.

I see now how the kv parameter to grok works.  Clever, but also very specialised :-)

**Comment by jaqx0r** [comment]
For the *specific* example you provided the easiest way to do this would be:

   /[?&]id=([^&].*)(&|$)/ {
     id = $1
   }

   /[?&]source=([^&].*)(&|$)/ {
     source = $1
   }

   /[?&]foo=([^&].*)(&|$)/ {
     foo = $1
   }

andd so on for each key, but yo uhave to know the keys in advance.

**Comment by jaqx0r** [comment]
I've figured out an approach to use in the VM to do the repeated matches on unknown keys, but it is very complex and is going to make the mtail programme complex as well to use it.  I don't like the way it looks nor the performance implication of the mtail VM if it's built.  I'd much rather have programmers explicitly list the keys they expect to see in their mtail programmes.

https://groups.google.com/d/msg/mtail-users/2hIi_qpIm2Y/wdjOXEa6CwAJ is another user asking for a similar feature.

Again I recommended this pattern instead of implementing the feature.
